webui/react/src/services/stream/stream.ts

```diff
@@ -0,0 +1,173 @@
+import { forEach, reduce, trimEnd } from 'lodash';
+
+import { decode_keys, KeyCache } from './keyCache';
+import { StreamSpec } from './projects';
+
+import { Streamable, StreamEntityMap } from '.';
+
+// About 60 seconds of auto-retry.
+const backoffs = [0, 1, 2, 4, 8, 10, 10, 10, 15];
+export const sleep = (s: number): Promise<unknown> => new Promise((r) => setTimeout(r, 1000 * s));
+
+type Subscription = {
+    keyCache: KeyCache,
+    spec: StreamSpec
+}
+
+export class Stream {
+
+    readonly #wsUrl: string;
+    #ws: WebSocket | undefined = undefined;
+    #retries: number = 0;
+    #numSyncs: number = 0;
+    #closedByClient: boolean = false;
+
+    #subs: Array<Record<Streamable, Subscription>> = [];
+    #curSub: Record<Streamable, Subscription> | undefined;
+
+    // syncSent updates when msg sent to the stream
+    #syncSent: string | undefined = undefined;
+    // syncStarted updates when recieving msg of {sync_id, complated: false}
+    #syncStarted: string | undefined = undefined;
+    // syncStarted updates when recieving msg of {sync_id, complated: true}
+    #syncComplete: string | undefined = undefined;
+
+    //callbacks
+    #onUpsert: (m: Record<string, any>) => void;
+    #onDelete: (s: Streamable, a: Array<number>) => void;
+
+    constructor(wsUrl: string, onUpsert: (m: Record<string, any>) => void, onDelete: (s: Streamable, a: Array<number>) => void) {
+        this.#wsUrl = wsUrl;
+        this.#onUpsert = onUpsert;
+        this.#onDelete = onDelete;
+        this.#connect();
+    }
+
+    #connect() {
+        this.#ws = new WebSocket(this.#wsUrl);
+        this.#ws.onopen = () => {
+            console.log('Streaming websocket opened!');
+            this.#advanceSubscription();
+        };
+        this.#ws.onerror = async (err) => {
+            console.log('Streaming websocket errored: ', err);
+            await this.#retry();
+        };
+
+        this.#ws.onclose = async () => {
+            if (!this.#closedByClient) await this.#retry();
```

Comment:
Also... this might be radical, but I think promises and async/await code make writing code like this _very_ difficult.  If you are interested I can teach you a pattern that I find easier for me.  You'll have to decide for yourself if you like it.

Reply:
Thanks, I'd love to hear more about it